### PR TITLE
MapService refactoring. Make partition destroy to be called from the partition thread

### PR DESCRIPTION
**Pull request by toktarev**
To be done:

1) Some refactoring
2) Make CleanPartitions in map.Destory to be called from the PartitionThread.

closes https://github.com/hazelcast/hazelcast/issues/4331

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java [hunk 1]
```diff
@@ -171,10 +180,29 @@ public void flushMaps() {
...
     public void destroyMap(String mapName) {
         final PartitionContainer[] containers = partitionContainers;
+        final List<Future> futures = new ArrayList<Future>(containers.length);
         for (PartitionContainer container : containers) {
             if (container != null) {
-                container.destroyMap(mapName);
+                int partitionId = container.getPartitionId();
+                Address partitionOwner = nodeEngine.getPartitionService().getPartitionOwner(partitionId);
+
+                if ((partitionOwner != null) && (partitionOwner.equals(nodeEngine.getThisAddress()))) {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/MapServiceContextImpl.java]: You can also use:
```
InternalPartition partition = nodeEngine.getPartitionService().getPartition(partitionId);
 if (partition.isLocal()) {

 }
```

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapPartitionDestroyOperation.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        return true;
+    }
+
+    @Override
+    public Object getResponse() {
+        return partitionContainer.getPartitionId();
+    }
+
+    @Override
+    protected void writeInternal(ObjectDataOutput out) throws IOException {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapPartitionDestroyOperation.java]: Can you put some comment why this method is empty for the future viewer?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapPartitionDestroyOperation.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    }
+
+    @Override
+    protected void readInternal(ObjectDataInput in) throws IOException {
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapPartitionDestroyOperation.java]: same here

File: hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import java.util.concurrent.BrokenBarrierException;
+
+import com.hazelcast.test.HazelcastParallelClassRunner;
+
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category(QuickTest.class)
+public class MapPutDestroyTest extends HazelcastTestSupport {
+    @Test
+    public void testMapPutDestroy() {
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java]: can you rename this test, currently it is hard to understand what it is testing
new name maybe something like `testConcurrentPutDestroy_doesNotCauseNPE `

File: hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        try {
+            t2.join();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java]: This test need assertions also. Ideally, i expect an always failing test without this PR.

Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java]: >> i expect an
I wouldn't expect it here - quite hardly reproduceable race condition

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapPartitionDestroyOperation.java (hunk 2 above)
Comment (toktarev) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapPartitionDestroyOperation.java]: Done

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapPartitionDestroyOperation.java (hunk 3 above)
Comment (toktarev) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapPartitionDestroyOperation.java]: Done

File: hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java (hunk 5 above)
Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java]: Out it to Nightly tests and added assertions

File: hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.hazelcast.test.HazelcastParallelClassRunner;
+
+import static junit.framework.TestCase.assertNull;
+
+
+@RunWith(HazelcastParallelClassRunner.class)
+@Category(NightlyTest.class)
+public class MapPutDestroyTest extends HazelcastTestSupport {
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java]: This test mostly fails on my local.
```
@Test
    public void testConcurrentPutDestroy_doesNotCauseNPE() throws Exception {
        TestHazelcastInstanceFactory factory = createHazelcastInstanceFactory(3);
        final HazelcastInstance instance = factory.newHazelcastInstance();
        final HazelcastInstance instance2 = factory.newHazelcastInstance();
        final HazelcastInstance instance3 = factory.newHazelcastInstance();
        final String mapName = randomString();
        final AtomicReference<Exception> error = new AtomicReference<Exception>(null);

        final AtomicBoolean stop = new AtomicBoolean();

        Thread t1 = new Thread(
                new Runnable() {
                    @Override
                    public void run() {
                        try {
                            while (!stop.get()) {
                                IMap<Object, Object> map = instance.getMap(mapName);
                                map.put(System.currentTimeMillis(), Boolean.TRUE);
                            }
                        } catch (Exception e) {
                            error.set(e);
                        }
                    }
                }
        );

        Thread t2 = new Thread(
                new Runnable() {
                    @Override
                    public void run() {
                        try {
                            while (!stop.get()) {
                                IMap<Object, Object> map = instance.getMap(mapName);
                                map.destroy();
                            }
                        } catch (Exception e) {
                            error.set(e);
                        }
                    }
                }
        );

        t1.start();
        t2.start();

        sleepSeconds(10);
        stop.set(true);

        try {
            t1.join();
        } catch (Throwable e) {
            e.printStackTrace();
        }

        try {
            t2.join();
        } catch (Throwable e) {
            e.printStackTrace();
        }


        Exception object = error.get();
        if (object instanceof NullPointerException) {
            fail();
        }

    }
```

Comment (toktarev) [inline hazelcast/src/test/java/com/hazelcast/map/MapPutDestroyTest.java]: Migrated it

**Comment by ahmetmircik** [comment]
:+1:  seems mergeable when PR builder is ok. @gurbuzali ?

**Comment by gurbuzali** [comment]
verify

**Pull request merged**